File: frontend-mobile-flutter/lib/services/workflow_execution_poll.dart
import '../api/workflows_api.dart';
import '../models/workflow.dart';

class WorkflowPollResult {
  WorkflowPollResult({
    required this.status,
    required this.savedResumes,
    required this.stepLogs,
    this.errorMessage,
  });

  final String status;
  final List<SavedResumeSummary> savedResumes;
  final List<WorkflowStepLog> stepLogs;
  final String? errorMessage;
}

class WorkflowExecutionPoll {
  WorkflowExecutionPoll(this._api);

  final WorkflowsApi _api;

  Future<WorkflowPollResult> poll(
    String groupId, {
    void Function(List<WorkflowStepLog> logs, double? progress)? onProgress,
    int maxAttempts = 600,
    Duration interval = const Duration(seconds: 1),
  }) async {
    for (var attempt = 0; attempt < maxAttempts; attempt++) {
      final data = await _api.getExecutionLogs(groupId);
      final rawLogs = data['stepLogs'] as List<dynamic>? ?? [];
      final logs = rawLogs
          .whereType<Map>()
          .map((e) => WorkflowStepLog.fromJson(Map<String, dynamic>.from(e)))
          .toList();
      final progress = (data['progress'] as num?)?.toDouble();
      onProgress?.call(logs, progress);

      final status = data['status'] as String?;
      if (status == 'completed') {
        final saved = (data['savedResumes'] as List<dynamic>? ?? [])
            .whereType<Map>()
            .map((e) => SavedResumeSummary.fromJson(Map<String, dynamic>.from(e)))
            .toList();
        return WorkflowPollResult(
          status: 'completed',
          savedResumes: saved,
          stepLogs: logs,
        );
      }
      if (status == 'failed') {
        return WorkflowPollResult(
          status: 'failed',
          savedResumes: const [],
          stepLogs: logs,
          errorMessage: data['errorMessage'] as String? ?? '执行失败',
        );
      }
      await Future<void>.delayed(interval);
    }
    return WorkflowPollResult(
      status: 'failed',
      savedResumes: const [],
      stepLogs: const [],
      errorMessage: '执行超时',
    );
  }
}

File: frontend-mobile-expo/src/utils/workflowExecutionPoll.ts
import { workflowsApi } from '@/api/workflows'

export interface SavedResumeSummary {
  id: number
  title: string
  templateId: string
  lang: string
}

export interface WorkflowStepLog {
  stepKey: string
  status: string
  message?: string
}

export interface WorkflowPollResult {
  status: 'completed' | 'failed'
  savedResumes: SavedResumeSummary[]
  errorMessage?: string
  stepLogs: WorkflowStepLog[]
}

function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export async function pollWorkflowExecution(
  groupId: string,
  onProgress?: (stepLogs: WorkflowStepLog[], progress?: number) => void,
): Promise<WorkflowPollResult> {
  const maxAttempts = 600

  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    const res = await workflowsApi.getExecutionLogs(groupId)
    const data = res.data as {
      status?: string
      savedResumes?: SavedResumeSummary[]
      errorMessage?: string
      stepLogs?: WorkflowStepLog[]
      progress?: number
    } | undefined

    if (data?.stepLogs) {
      onProgress?.(data.stepLogs, data.progress)
    }

    if (data?.status === 'completed') {
      return {
        status: 'completed',
        savedResumes: data.savedResumes ?? [],
        stepLogs: data.stepLogs ?? [],
      }
    }

    if (data?.status === 'failed') {
      return {
        status: 'failed',
        savedResumes: [],
        errorMessage: data.errorMessage ?? '执行失败',
        stepLogs: data.stepLogs ?? [],
      }
    }

    await sleep(2000)
  }

  throw new Error('执行超时，请稍后在执行记录中查看结果')
}
